src/CSE.cpp

```diff
@@ -61,9 +61,12 @@ bool should_extract(const Expr &e) {
     }
 
     if (const Call *a = e.as<Call>()) {
-        if (!a->is_pure()) {
-            //non-pure Calls may have side-effects, thus
-            //may not be re-ordered or reduced in number.
+        if (!a->is_pure() && (a->call_type != Call::Halide)) {
```

Comment:
The problem is that they're not pure inside the loop nest of an update definition for the same Func. But they can only change values across stmts - within a single Expr they're pure.